gapic/schema/api.py

```diff
@@ -526,24 +532,73 @@ def _get_methods(self,
                         'metadata type.',
                     )
                 lro = wrappers.OperationInfo(
-                    response_type=self.api_messages[address.resolve(
+                    response_type=self.api_messages[service_address.resolve(
                         op.response_type,
                     )],
-                    metadata_type=self.api_messages[address.resolve(
+                    metadata_type=self.api_messages[service_address.resolve(
                         op.metadata_type,
                     )],
                 )
 
+            # If we got a gRPC service config, get the appropriate retry
+            # and timeout information from it.
+            retry = None
+            timeout = None
+            if self.opts.retry:
+                # This object should be a dictionary that conforms to the
+                # gRPC service config proto:
+                #   Repo: https://github.com/grpc/grpc-proto/
+                #   Filename: grpc/service_config/service_config.proto
+                #
+                # We only care about a small piece, so we are just leaving
+                # it as a dictionary and parsing accordingly.
+                for mc in self.opts.retry.get('methodConfig', []):
+                    # Does this method config apply to us?
+                    # The gRPC service config uses a repeated `name` field
+                    # with a particular format, which we match against.
+                    selector = {
+                        'service': '{package}.{service_name}'.format(
+                            package='.'.join(service_address.package),
+                            service_name=service_address.name,
+                        ),
+                        'method': meth_pb.name,
+                    }
+                    if selector not in mc.get('name', []):
+                        continue
+
+                    # Set the retry and timeout according to this rule.
+                    if mc.get('timeout'):
+                        timeout = self._to_float(mc['timeout'])
+                    r = mc.get('retryPolicy', {})
+                    retry = wrappers.RetryInfo(
+                        max_attempts=r.get('maxAttempts', 0),
+                        initial_backoff=self._to_float(
+                            r.get('initialBackoff', '0s'),
+                        ),
+                        max_backoff=self._to_float(
+                            r.get('maxBackoff', '0s'),
+                        ),
+                        backoff_multiplier=r.get('backoffMultiplier', 0.0),
```

Comment:
Outside the scope of the review, but does this imply that the default values are provided by the default generated service config?

Reply:
The template declines to use zero values. In theory, all of these fields are required, but I do not trust that to actually happen.